src/Propel/Runtime/ActiveQuery/QueryExecutor/AbstractQueryExecutor.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+/**
+ * MIT License. This file is part of the Propel package.
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Propel\Runtime\ActiveQuery\QueryExecutor;
+
+use Propel\Runtime\ActiveQuery\Criteria;
+use Propel\Runtime\ActiveQuery\SqlBuilder\PreparedStatementDto;
+use Propel\Runtime\Connection\ConnectionInterface;
+use Propel\Runtime\Connection\ConnectionWrapper;
+use Propel\Runtime\Connection\StatementWrapper;
+use Propel\Runtime\Propel;
+use Propel\Runtime\ServiceContainer\ServiceContainerInterface;
+use Throwable;
+
+abstract class AbstractQueryExecutor
+{
+    /**
+     * Indicates if the executor performs writes.
+     * Defaults to true, but is overridden by subclasses.
+     *
+     * @see AbstractQueryExecutor::getConnection()
+     *
+     * @var bool
+     */
+    protected const NEEDS_WRITE_CONNECTION = true;
+
+    /**
+     * @var \Propel\Runtime\Connection\ConnectionInterface
+     */
+    protected $con;
+
+    /**
+     * @var \Propel\Runtime\Adapter\SqlAdapterInterface
+     */
+    protected $adapter;
+
+    /**
+     * @var \Propel\Runtime\ActiveQuery\Criteria
+     */
+    protected $criteria;
+
+    /**
+     * @var \Propel\Runtime\Map\DatabaseMap
+     */
+    protected $dbMap;
+
+    /**
+     * @param \Propel\Runtime\ActiveQuery\Criteria $criteria
+     * @param \Propel\Runtime\Connection\ConnectionInterface|null $con
+     */
+    public function __construct(Criteria $criteria, ?ConnectionInterface $con = null)
+    {
+        $this->criteria = $criteria;
+
+        $dbName = $criteria->getDbName();
+        $serviceContainer = Propel::getServiceContainer();
+
+        $this->con = ($con) ? $con : $this->getConnection($serviceContainer, $dbName, static::NEEDS_WRITE_CONNECTION);
+
+        /** @var \Propel\Runtime\Adapter\SqlAdapterInterface $adapter */
+        $adapter = $serviceContainer->getAdapter($dbName);
+        $this->adapter = $adapter;
+
+        $this->dbMap = $serviceContainer->getDatabaseMap($dbName);
+    }
+
+    /**
+     * Retrieves a read or write connection from the service container.
+     *
+     * @param \Propel\Runtime\ServiceContainer\ServiceContainerInterface $sc
+     * @param string $dbName
+     * @param bool $getWritableConnection
+     *
+     * @return \Propel\Runtime\Connection\ConnectionInterface
+     */
+    protected function getConnection(ServiceContainerInterface $sc, string $dbName, bool $getWritableConnection = true): ConnectionInterface
```

Comment:
(⚠strong opinions following)
getters and/or setters usually hint at a broken design, leaking implementation details and making maintenance tricky as users of the Propel library _will_ start to depend on the specific exposed implementation.

Also: this `getConnection` function doesn't reference `$this` at all. Should it really be a method of this class?

Reply:
This isn't really a getter, the name was poorly chosen (changed it to `retrieveConnection`). I would have liked to make it a static method, but those are frowned upon in the project